services/client/storage_helper.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"errors"
+
+	"github.com/golang/protobuf/proto"
+	schema "github.com/m3db/m3cluster/generated/proto/placement"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/proto/util"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+)
+
+var (
+	errInvalidProtoForSinglePlacement    = errors.New("invalid proto for single placement")
+	errInvalidProtoForPlacementSnapshots = errors.New("invalid proto for placement snapshots")
+	errNoPlacementInTheSnapshots         = errors.New("not placement in the snapshots")
+)
+
+type placementStorageHelper interface {
+	// Placement retrieves the placement stored on kv.Store.
+	Placement(store kv.Store, key string) (services.Placement, int, error)
```

Comment:
because the PlacementStorage interface takes in a different service id in the methods signature, so I have to create a different store each time by looking at the sid, I could create a helper every time a new kv.Store is created and ref the store in the helper, but felt like an over kill since this is an internal interface.

Reply:
So the same placement storage can generate placements for different services presumably? I guess this is fine if that's the case.

Alternatively I wonder if it's cleaner if the placement storage takes in a service id at construction so handling different services from the same storage is a non-issue.

Reply:
yes current placement storage interface can access placement of different serviceIDs, that would make the placementStorageHelper uglier but the use of placement storage interface easier since the user don't need to create a storage per service ID. Since the helper is an internal interface I think this is ok.